### DISCUSSION TITLE
making sources of complex shapes

### DISCUSSION THREAD

**Discussion by yurivict** [NanoComp/meep#2905] [Q&A]
I modified the source of the example bent-waveguide.ipynb to have 6 vertices:
[bent-waveguide-1.ipynb](https://freebsd.org/~yuri/bent-waveguide-1.ipynb)

Instead of drawing a polygon with 6 vertices it draws a smaller square:
![image](https://github.com/user-attachments/assets/d038f43f-e929-49af-a6ee-efbf8a875bdf)

This square is obviously smaller than 2x2 that was specified in the modified example.

Why is the displayed shape smaller than specified in the vertices array, and why doesn't it have 6 vertices?

Thank you.

**Comment by stevengj** [comment]
`vertices` is an argument of `mp.Prism`, not `mp.Volume`.

**Comment by yurivict** [comment]
@stevengj 

It says [here](https://github.com/NanoComp/meep/blob/master/python/simulation.py#L409) in the description of the class mp.Volume:
> Alternatively, you can specify a list of `Vector3` vertices using the `vertices` parameter. The `center` and `size` will automatically be computed from this list.

And the constructor of the mp.Volume class code accepts the named 'volume' argument.

**Reply by stevengj** [reply]
It still doesn't make a polygon, just a box centered at the mean of the vertices.  A `Volume` is *always a box* aligned with the cartesian axes.

**Reply by yurivict** [reply]
Ok, but the size of the box doesn't correspond to the coordinates in the 'vertices' array: https://github.com/NanoComp/meep/issues/2906

**Reply by yurivict** [reply]
@stevengj 
If mp.Volume doesn't make a polygon, and only makes a box, how to create a source that has a polygon as its shape?
The mp.Source class only has the 'volume' argument that can only accept mp.Volume instances.
And how to create a 3D source of a complex shape?
Are there any examples for this?

**Comment by stevengj** [comment]
To make a source that covers a complicated shape (e.g. a polygon) that cannot easily be composed from Cartesian boxes, you should

1. Define the `volume` (or `center` and `size`) of the source to be a *bounding box* that encloses the intended source region.
2. Set the `amp_func` of the source to be a Python function that returns whatever source you want (e.g. a source that is 1 in a polygon and zero outside).

(Alternatively, you can add a large number of source "boxes" that fill your desired shape.)
